File: GoLang/DS/List/SLinkList.go
package main

import "fmt"

type Node struct {
	val  interface{}
	next *Node
}

func newNode(val interface{}) *Node {
	return &Node{val, nil}
}

type LinkedList struct {
	head   *Node
	length int
}

func initLinkedList() *LinkedList {
	return &LinkedList{}
}

func (ll *LinkedList) prepend(data interface{}) {
	n := newNode(data)
	n.next = ll.head
	ll.head = n
	ll.length++
}

func (ll *LinkedList) append(data interface{}) {
	n := newNode(data)
	if ll.head == nil {
		ll.head = n
		ll.length++
		return
	}
	curr := ll.head
	for ; curr.next != nil; curr = curr.next {

	}
	curr.next = n
	ll.length++
}

func (ll *LinkedList) removeFront() interface{} {
	if ll.head == nil {
		return nil
	}

	curr := ll.head
	ll.head = ll.head.next
	ll.length--
	return curr.val
}

func (ll *LinkedList) removeBack() interface{} {
	if ll.head == nil {
		return nil
	}

	if ll.head.next == nil {
		curr := ll.head
		ll.head = nil
		ll.length--
		return curr.val
	}

	curr := ll.head
	for ; curr.next != nil; curr = curr.next {

	}

	ll.length--
	retVal := curr.next.val
	curr.next = nil
	return retVal
}

func (ll *LinkedList) size() int {
	return ll.length
}

func (ll *LinkedList) display() {
	curr := ll.head
	for ; curr != nil; curr = curr.next {
		fmt.Print(curr.val, " ")
	}
	fmt.Println()
}

func (ll *LinkedList) reverse() {
	if ll.head == nil {
		return
	}

	if ll.head.next == nil {
		return
	}

	var curr, rev, temp *Node
	curr = ll.head
	rev = nil
	temp = nil

	for curr != nil {
		temp = curr.next
		curr.next = rev
		rev = curr
		curr = temp
	}

	ll.head = rev
}

func (ll *LinkedList) insertAtN(index int, data interface{}) {
	if index < 0 || index > ll.length {
		fmt.Println("Insert: Index Out of Bound")
		return
	}

	if index == 0 {
		ll.prepend(data)
		return
	}

	if index == ll.length {
		ll.append(data)
		return
	}
	curr := ll.head

	for i := 0; i < index; i++ {
		curr = curr.next
	}
	n := newNode(data)
	n.next = curr.next
	curr.next = n
	ll.length++
	return
}

func (ll *LinkedList) removeAtN(index int) {
	if index < 0 || index > ll.length-1 {
		fmt.Println("Insert: Index Out of Bound")
		return
	}

	if index == 0 {
		ll.removeFront()
		return
	}

	if index == ll.length-1 {
		ll.removeBack()
		return
	}
	curr := ll.head
	var prev *Node

	for i := 0; i < index; i++ {
		prev = curr
		curr = curr.next
	}
	prev.next = curr.next
	ll.length--
}

func (ll *LinkedList) contains(data interface{}) bool {
	if ll.head == nil {
		return false
	}

	curr := ll.head
	for curr != nil {
		if curr.val == data {
			return true
		}
		curr = curr.next
	}
	return false
}

func main() {
	list := initLinkedList()

	for i := 0; i < 10; i++ {
		list.append(i)
	}

	list.display()
	list.reverse()
	list.display()
}

File: DsAndAlgo/Scala/DsAndAlgo/src/main/scala/SLinkList.scala
package DS.List

import scala.util.control.Breaks

class  SLinkList[T]  {
  var head:SLinkNode[T] = null;
  var curr:SLinkNode[T] = null;
  var len:Int = 0;

  /* Helper APIs */
  def isEmpty(): Boolean = {
    if (head == null) {
      true
    } else {
      false
    }
  }

  def size(): Int = {
    this.len
  }

  def top():T = {
    assert(head != null)
    getData(0)
  }

  def getData(pos: Int): T = {
    assert(pos >= 0 && pos < len && head != null);
    var temp:SLinkNode[T] = head;
    var idx:Int = 0;

    while (idx != pos) {
      temp = temp.getNext;
    }

    assert(temp != null);
    temp.data;
  }

  def getIdx(data:T):Int = {
    assert(head != null);
    var temp:SLinkNode[T] = head;
    var idx = 0;

    while(temp != null) {
      if ((temp.getData()).equals(data)) {
        return idx;
      } else {
        idx = idx + 1;
      }
    }

    assert(temp == null);
    idx;
  }

  /* Iterative API */

  def cur(): T = {
    this.curr.getData;
  }

  def next():Unit  = {
    this.curr = this.curr.getNext;
  }

  def move(pos:Int) :Unit = {
    assert(pos >= 0 && pos < len);
    this.curr = this.head;
    var idx:Int = 0;

    while (idx != pos) {
      this.curr = this.curr.getNext;
    }

  }

  def reset():Unit = {
    this.curr = this.head;
  }

  def last:Unit = {
    this.curr = this.head;

    while(this.curr.getNext != null) {
      this.curr = this.curr.getNext;
    }

  }

  /* Search API */

  def contains(data:T):Boolean = {
    var temp:SLinkNode[T] = this.head;

    while(temp != null) {
      if ((temp.getData())equals()) {
        return true;
      } else {
          temp = temp.getNext;
      }
    }

    false;
  }

  /* Insert APIs */
  def insert(data:T):Unit = {
    var newNode  = new SLinkNode[T](data, null)

    head match {
      case null => {
        head = newNode
        curr = head
      }

      case _ => {
        newNode.next = curr.next
        curr.next = newNode;
        curr = newNode;
      }
    }
    len = len + 1;
  }

  def insert(data:T, pos:Int): Unit = {
    var newNode = new SLinkNode[T](data, null)
    assert(pos >= 0 && pos < len)

    pos match {
      case 0 => {
        newNode.next = head;
        head = newNode
      }
      case _ => {
        var temp = head
        var idx = 0;
        while(idx < pos) {
          idx = idx + 1
          temp = temp.next;
        }
        newNode.next = temp.next
        temp.next= newNode
      }
    }

    len = len + 1
  }

  def append(data:T): Unit = {
      insert(data, len - 1)
  }

  def prepend(data:T): Unit = {
      insert(data, 0)
  }

  /* remove APIs */
  def remove(data:T):Unit = {
    var prev = head;
    var temp = head;

    val loopBreak = new Breaks;
    assert(data != null)
    loopBreak.breakable {
      while (temp != null) {
        if (temp.data.equals(data)) {
          if (temp equals (prev)) {
            head = temp.next
          } else {
            prev.next = temp.next;
          }
          if (curr equals (temp)) {
            curr = temp.next
          }
          len = len - 1
          loopBreak.break()
        } else {
          prev = temp
          temp = temp.next
        }
      }
    }
  }


  def pop(data:T) = {
        assert(head != null);
        if (curr.equals(head)) {
          curr = head.next
        }
        head = head.next
        len = len - 1
  }

  def dequeue(data:T) :Unit = {
    assert(head != null)
    var data = getData(len - 1)
    remove(data)
  }

  def remove(): Unit = {
      var data = curr.data
      remove(data)
  }

  /* Misc */
  def reverse():Unit = {
    var curNode:SLinkNode[T] = null
    var revList :SLinkNode[T] = null
    var nextNode :SLinkNode[T] = null;

    curNode = head;
    while (curNode != null) {
      nextNode = curNode.next
      curNode.next = revList;
      revList = curNode
      curNode = nextNode
    }

    head = revList;
    curr = head
  }

  override def toString(): String = {
    var strbuf = new StringBuilder("")
    var itrNode = head
    while (itrNode != null) {
      strbuf = strbuf.append(new StringBuilder(itrNode.data.toString + " "))
      itrNode = itrNode.next
    }
    var str = strbuf.toString()
    str
  }

  def clear():Unit = {
    while (head != null) {
      head = head.next
    }
    curr = null
    len = 0
  }
}
